rpc/dapserver/server.go

```diff
@@ -0,0 +1,108 @@
+package dapserver
+
+import (
+	"bufio"
+	"context"
+	"errors"
+	"io"
+	"io/ioutil"
+	"log"
+
+	"github.com/chzyer/readline"
+	dap "github.com/google/go-dap"
+	"github.com/openllb/hlb/codegen"
+	"golang.org/x/sync/errgroup"
+)
+
+type Server struct {
+	dbgr codegen.Debugger
+}
+
+func New(dbgr codegen.Debugger) *Server {
+	return &Server{dbgr}
+}
+
+func (s *Server) Listen(ctx context.Context, output, stdin io.Reader, stdout io.Writer) error {
+	ctx, cancel := context.WithCancel(ctx)
+	cancelableStdin := readline.NewCancelableStdin(stdin)
+	session := Session{
+		dbgr: s.dbgr,
+		rw: bufio.NewReadWriter(
+			bufio.NewReader(cancelableStdin),
+			bufio.NewWriter(stdout),
+		),
+		cancel:            cancel,
+		sendQueue:         make(chan dap.Message),
+		caps:              make(map[Capability]struct{}),
+		sourcesHandles:    newHandlesMap(),
+		variablesHandles:  newHandlesMap(),
+		stackFrameHandles: newHandlesMap(),
+	}
+
+	g, ctx := errgroup.WithContext(ctx)
+
+	g.Go(func() error {
+		return session.sendFromQueue(ctx)
+	})
+
+	if output == nil {
+		g.Go(func() error {
+			<-ctx.Done()
+			return cancelableStdin.Close()
+		})
+	} else {
+		g.Go(func() error {
+			defer cancelableStdin.Close()
+
+			scanner := bufio.NewScanner(output)
+			for scanner.Scan() {
+				session.send(&dap.OutputEvent{
+					Event: newEvent("output"),
+					Body: dap.OutputEventBody{
+						Category: "stdout",
+						Output:   scanner.Text() + "\n",
+					},
+				})
+				select {
+				case <-ctx.Done():
+					return nil
+				default:
+				}
+			}
+
+			return scanner.Err()
+		})
+	}
+
+	// f, err := os.Create("/tmp/hlb-dapserver.log")
+	// if err != nil {
+	// 	panic(err)
+	// }
+	// defer f.Close()
+	// log.SetOutput(f)
```

Comment:
Remove or clean up?

Reply:
This is mostly for debugging purposes, I think eventually when dapserver can run on a unix socket / tcp and allow multiple clients it may be refactored from `hlb run --dap` to `hlb dapserver` and log output will be a flag to that command.

Multiple `codegen.Debugger` clients is necessary because IDE DAP doesn't provide a good way to do exec. My current idea is that the IDE integrated terminal is connected to the TUI debugger, so controlling the debugger can be done from both UI or terminal (`exec` being only available on the terminal).